src/asm/warning.c

```diff
@@ -0,0 +1,244 @@
+
+#include <limits.h>
+#include <stdarg.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+
+#include "asm/fstack.h"
+#include "asm/main.h"
+#include "asm/warning.h"
+
+#include "extern/err.h"
+
+unsigned int nbErrors; /* = 0 */
+
+enum WarningState {
+	WARNING_DEFAULT,
+	WARNING_DISABLED,
+	WARNING_ENABLED,
+	WARNING_ERROR
+};
+
+static enum WarningState const defaultWarnings[NB_WARNINGS] = {
+	WARNING_ENABLED,  /* User warnings */
+	WARNING_DISABLED, /* Obsolete things */
+	WARNING_DISABLED, /* Invalid args to builtins */
+	WARNING_DISABLED, /* Constants too large */
+	WARNING_DISABLED, /* Shifting undefined behavior */
+	WARNING_DISABLED, /* Division undefined behavior */
+	WARNING_DISABLED, /* Empty entry in `db`, `dw` or `dl` */
+	WARNING_DISABLED, /* String too long for internal buffers */
+};
+
+static enum WarningState warningStates[NB_WARNINGS];
+
+static bool warningsAreErrors; /* Set if `-Werror` was specified */
+
+static enum WarningState warningState(enum WarningID id)
+{
+	/* Check if warnings are globally disabled */
+	if (!CurrentOptions.warnings)
+		return WARNING_DISABLED;
+
+	/* Get the actual state */
+	enum WarningState state = warningStates[id];
+
+	if (state == WARNING_DEFAULT)
+		/* The state isn't set, grab its default state */
+		state = defaultWarnings[id];
+
+	if (warningsAreErrors && state == WARNING_ENABLED)
+		state = WARNING_ERROR;
+
+	return state;
+}
+
+static char const *warningFlags[NB_WARNINGS_ALL] = {
+	"user",
+	"obsolete",
+	"builtin-args",
+	"large-constant",
+	"shift",
+	"div",
+	"empty-entry",
+	"long-string",
+
+	/* Meta warnings */
+	"all",
+	"extra",
+	"everything" /* Especially useful for testing */
+};
+
+enum MetaWarningCommand {
+	META_WARNING_DONE = NB_WARNINGS
+};
+
+/* Warnings that probably indicate an error */
+static uint8_t const _wallCommands[] = {
+	WARNING_USER,
+	WARNING_BUILTIN_ARG,
+	WARNING_LARGE_CONSTANT,
+	WARNING_EMPTY_ENTRY,
+	WARNING_LONG_STR,
+	META_WARNING_DONE
+};
+
+/* Warnings that are less likely to indicate an error */
+static uint8_t const _wextraCommands[] = {
+	WARNING_OBSOLETE,
+	META_WARNING_DONE
+};
+
+/* Literally everything. Notably useful for testing */
+static uint8_t const _weverythingCommands[] = {
+	WARNING_USER,
+	WARNING_OBSOLETE,
+	WARNING_BUILTIN_ARG,
+	WARNING_LARGE_CONSTANT,
+	WARNING_SHIFT,
+	WARNING_DIV,
+	WARNING_EMPTY_ENTRY,
+	WARNING_LONG_STR,
+	META_WARNING_DONE
+};
+
+static uint8_t const *metaWarningCommands[NB_META_WARNINGS] = {
+	_wallCommands,
+	_wextraCommands,
+	_weverythingCommands
+};
+
+void processWarningFlag(char const *flag)
+{
+	static bool setError = false;
+
+	/* First, try to match against a "meta" warning */
+	for (enum WarningID id = NB_WARNINGS; id < NB_WARNINGS_ALL; id++) {
+		/* TODO: improve the matching performance? */
+		if (!strcmp(flag, warningFlags[id])) {
+			/* We got a match! */
+			uint8_t const *ptr =
+					metaWarningCommands[id - NB_WARNINGS];
+
+			for (;;) {
+				switch (*ptr) {
```

Comment:
A `switch` with just one `case` and a `default` label? Looks like it should've been an `if` to me. Particularly when the matching `case` just returns — a single `if (*ptr == META_WARNING_DONE) return;` line does the trick while being a lot simpler to read.

Reply:
The switch is there for future-proofing, since it's an enum of actions. Also, the one-liner version is made impossible by checkpatch.

Reply:
Will checkpatch complain about anything but the lack of a newline after the condition?

Reply:
```
issotm@sheik-kitty rgbds% make checkpatch
[*] Analyzing commit 'f13adf9172f8eea7327e095d87600390eca7fab9'
src/asm/warning.c:125: ERROR: trailing statements should be on next line
+				if (*ptr == META_WARNING_DONE) return;
```

Reply:
So the two-line equivalent works... as for the "just in case" `switch`, [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) (and it's easy to reintroduce if needed).

Reply:
Fixed.